src/geometry/manifold/real_instances.lean

```diff
@@ -0,0 +1,375 @@
+/-
+Copyright (c) 2019 Sébastien Gouëzel. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sébastien Gouëzel
+-/
+
+import geometry.manifold.smooth_manifold_with_corners linear_algebra.finite_dimensional
+
+/-!
+# Constructing examples of manifolds over ℝ
+
+We introduce the necessary bits to be able to define manifolds modelled over ℝ^n, boundaryless
+or with boundary or with corners. As a concrete example, we construct explicitly the manifold with
+boundary structure on the real interval [x, y].
+
+More specifically, we introduce
+* `euclidean_space n` for a model vector space of dimension `n`.
+* `model_with_corners ℝ (euclidean_space n) (euclidean_half_space n)` for the model space used
+to define `n`-dimensional real manifolds with boundary
+* `model_with_corners ℝ (euclidean_space n) (euclidean_quadrant n)` for the model space used
+to define `n`-dimensional real manifolds with corners
+
+## Implementation notes
+
+The manifold structure on the interval [x, y] = Icc x y requires the assumption `x < y`. We
+introduce a dummy class `lt_class x y` for this, to make such an assumption available to typeclass
+search. This should hopefully not be necessary any more in Lean 4.
+-/
+
+noncomputable theory
+open set
+
+/--
+The space ℝ^n. Note that the name is slightly misleading, as we only need a normed space
+structure on ℝ^n, but the one we use here is the sup norm and not the euclidean one -- this is not
+a problem for the manifold applications, but should probably be refactored at some point.
+-/
+def euclidean_space (n : ℕ) : Type := (fin n → ℝ)
+
+/--
+The half-space in ℝ^n, used to model manifolds with boundary. We only define it when `1 ≤ n`, as the
+definition only makes sense in this case.
+-/
+def euclidean_half_space (n : ℕ) [has_zero (fin n)] : Type :=
```

Comment:
Does this mean that `{pt}` is not a manifold with boundary? Can we find a definition that also works for `n = 0`?
(I don't know anything about this part of maths..., maybe this is just how it works. I was just assuming that any ordinary boundaryless manifold would also be a manifold with boundary.)

Reply:
I had initially a definition that worked also for `n=0`, as
```lean
def euclidean_half_space (n : ℕ) : Type := {x : euclidean_space n // ∀i:fin n, i.val = 0 → 0 ≤ x i}
```
But this definition is completely artificial, and it makes proofs more cumbersome. As one will never ever use manifolds with boundary for `n = 0` (since they don't really make sense), I decided to go for the more natural definition with the assumption `[has_zero (fin n)]`. As this is a type class assumption, it should never make statements more verbose to apply as it should always be checked automatically.

And yes, a manifold with boundary can usually be considered as a manifold with boundary, or even as a manifold with corners, by composing the charts with the exponential map along each coordinate, for instance. But this really has no interest in dimension 0, so I don't mind carrying around the assumption `[has_zero(fin n)]`.

A basic statement in all textbooks is that if `M` is a manifold with boundary of dimension `n`, then its boundary is a manifold without boundary of dimension `n-1`. This only makes sense for positive `n`...

Reply:
Ok, fair enough!